### PR TITLE
[MM-41347] : Advanced text editor

### PR DESCRIPTION
**Pull request by AshishDhama**
#### Summary
Advance Text Editor

#### Ticket Link
  Fixes https://mattermost.atlassian.net/browse/MM-33692

#### Related Pull Requests
NONE

#### Screenshots
|  before  |  after  |
|----|----|
| Screenshot 2022-05-09 at 3 46 34 PM | Screenshot 2022-05-09 at 3 45 32 PM |
| Screenshot 2022-05-09 at 3 46 22 PM | Screenshot 2022-05-09 at 3 45 41 PM |

#### Release Note
```release-note
* UI: Advanced Text editor with new shortcuts for preview, toggle formatter, open emoji picker, strike text, heading, Ordered list , Unordered list, Code)
```

### CODE REVIEW

**Comment by AshishDhama** [comment]
@esethna Although I still need to fix some cases, please take a look if you get time.

**Comment by esethna** [comment]
@AshishDhama thanks, looking good! A few things

1) Please ensure it's expanded by default for new and existing users (I created a new user and it was collapsed by default). If they don't want it they can collapse it and the state will be remembered. 
2) Hitting ENTER in the preview mode should probably send the messages and go back to editing mode in the empty text box cc// @matthewbirtch on thoughts here
3) The ... menu in the RHS gets stuck open, probably the same for the center channel at small widths
image
4) Pressing the buttons when the cursor is at the end of another word should probably add a space before the next word:
image
5) Looks like we still need support for the undo mechanism? Should we add in another PR? If so we should do it before this releases

**Comment by matthewbirtch** [comment]
> * Hitting ENTER in the preview mode should probably send the messages and go back to editing mode in the empty text box cc// @matthewbirtch on thoughts here

agree

> 6\. Input boxes are missing rounded corners in their unselected state, not sure if this was in the design @matthewbirtch?

Nope, definitely not as designed. It should maintain the same rounded corners in all states

@AshishDhama I'm not yet tagged as a reviewer, so I'm holding off. Also, this PR is marked as `Draft` state, so just add me as a reviewer when you think it's ready for a detailed UI review

**Comment by stylianosrigas** [comment]
Removing cloud test server for maintenance reason. Please feel free to add it again if needed.

**Comment by matthewbirtch** [comment]
> I will check these, but would suggest to add these in follow up pr. Rest I will work on.

Makes sense to me. Would you mind creating new tickets for those for follows ups?

**Review by esethna — APPROVED** [review]
Looks good! Approving to unblock merge,

In a follow-up PR we need to add the undo mechanism

**Comment by AshishDhama** [comment]
@michelengelen Hey if you could take a look to raise early red flags.

**Comment by AshishDhama** [comment]
@nevyangelova Thanks for the Awesome Head Start on Advance Text editor, Although work is still in progress but would be great if you can take look.
@mickmister @larkox I remember you guys were working on the same components. Please keep this in mind.

File: components/advance_text_editor/advance_text_editor.scss
Comment (nevyangelova) [reply components/advance_text_editor/advance_text_editor.scss:171]: the pseudo selector might not suffice as the class is removed on mouseup. Depends on the use case i guess. From a first look it seems that we wanna keep the 'active' style longer than just on mouse down

**Review by nevyangelova — CHANGES_REQUESTED** [review]
@AshishDhama Let's merge this as it is when the nit comments are fixed, but to keep in mind that we need to unify create_post, create_comment, create_post_advanced, create_comment_advanced when the feature is approved.

I understand now we are splitting them in this way so we can utilise the feature flag easier.

File: packages/mattermost-redux/src/selectors/entities/preferences.ts
Comment (AshishDhama) [reply packages/mattermost-redux/src/selectors/entities/preferences.ts:224]: `getIsAdvancedTexteditor` we are using both `is` and `get` personally I like `is` for Boolean so not changing this, and about `true` I know we don’t want it to keep it true, this was for testing purposes, there is no way to enable it on the Test server. Feature flag server branch wasn’t merged even now I want it to be true we reach a state where we are happy to merge this.

File: components/advance_text_editor/show_format/show_format.tsx
Comment (AshishDhama) [reply components/advance_text_editor/show_format/show_format.tsx]: It’s needed because we need this state when the editor is in preview state to show the button in active state.

File: components/advance_text_editor/toggle_formatting_bar/toggle_formatting_bar.tsx
Comment (AshishDhama) [reply components/advance_text_editor/toggle_formatting_bar/toggle_formatting_bar.tsx:37]: Same as earlier.

File: components/channel_view/channel_view.tsx
Comment (AshishDhama) [reply components/channel_view/channel_view.tsx:16]: Frankly speaking, naming it CreatePostAdvance or AdvancedCreatePost or whatever else doesn’t matter it’s temporary due to Feature Flag eventually it will be either AdvancedTextEditor or CreatePost only as we will delete duplicate once we feel all is okay with this component, and sooner we do it the better, that’s why I wasn’t in favor of feature flag but you insisted on it so here we are.

Comment (AshishDhama) [reply components/channel_view/channel_view.tsx:16]: Yeah, I have already created a ticket for that and will add these things to the ticket, and Todo in code, I have a request. Please try to refrain from naming suggestions until it’s a blunder because that’s a subjective topic. That discussion was going on before I was born and will keep going even after I die, so let’s not waste time on that topic. Thanks for the other suggestions.

File: components/channel_view/channel_view.tsx
Comment (AshishDhama) [reply components/channel_view/channel_view.tsx:37]: Subjective

File: components/threading/virtualized_thread_viewer/create_comment.tsx
Comment (AshishDhama) [reply components/threading/virtualized_thread_viewer/create_comment.tsx:75]: Subjective

File: components/threading/virtualized_thread_viewer/create_comment.tsx
Comment (AshishDhama) [reply components/threading/virtualized_thread_viewer/create_comment.tsx:40]: Subjective

File: components/keyboard_shortcuts/keyboard_shortcuts_sequence/keyboard_shortcuts.ts
Comment (AshishDhama) [reply components/keyboard_shortcuts/keyboard_shortcuts_sequence/keyboard_shortcuts.ts:249]: Yeah totally agree we can improve this, but I don’t think this is part of this ticket. If we try to fix whatever we feel is off, we will end up fixing something totally different altogether.

File: components/create_post_advance/create_post_advance.tsx
Comment (AshishDhama) [reply components/create_post_advance/create_post_advance.tsx]: Yeah, A problem for some other day. FYI, It’s part of the Web Platform OKR, and I will help them but let them lead this.

File: components/create_post_advance/create_post_advance.tsx
Comment (AshishDhama) [reply components/create_post_advance/create_post_advance.tsx]: Subjective

**Comment by stafot** [comment]
@jgilliam17 Shouldn't this pass the tests before we run the Cloud Test Server?

**Comment by jgilliam17** [comment]
@stafot This may not have been the best PR to give as an example of server not being created. 
The "blocker" check is failing on other PRs too and preventing server creation and merge. I've added a different one to our [thread](https://community-daily.mattermost.com/private-core/pl/smhzqqwf9idj8xrhkftffqieoc)

**Comment by stafot** [comment]
@jgilliam17 Agreed, I am investigating the root cause of this

**Comment by jgilliam17** [comment]
@AshishDhama few observations as I mentioned in the meeting earlier, apologies if some are duplicates. I can also split from this PR and file for follow up if necessary 
1. Unable to post by hitting Enter while in preview mode
2. If preview is ON and user hides the formatting bar, and we can't post using Enter, the experience is slightly broken as it's not visually apparent that the preview is enabled
3. Able to add emoji to draft post in preview mode
4. Able to remove attachments in preview mode
5. 3-dot menu is disconnected from `...` button in narrow widths
Screen Shot 2022-05-17 at 1 58 15 PM

6. 3-dot menu opacity is not correct when preview mode is active 
Screen Shot 2022-05-17 at 1 45 01 PM

7. Should we close 3-dot menu on outside click? Currently user has to click on the 3-dot button again to close, can't be dismissed in any other way

8. When in preview mode and clicking on formatting options buttons, a long string can be inadvertently added to the back of the draft text, user can accidentally miss any changes being added on preview screen
Screen Shot 2022-05-17 at 2 09 59 PM
9. The send button doesn't show/hide dynamically when changing the window size.
Screen Shot 2022-05-17 at 2 18 46 PM

**Comment by jgilliam17** [comment]
This is definitely going to break e2es. I haven't run the cycle yet, waiting for all green checks to trigger.

**Comment by jgilliam17** [comment]
Two more points:
1. 3-dot menu tucks under file icon (preview off)
Screen Shot 2022-05-17 at 3 23 06 PM

2. As @matthewbirtch previously mentioned, there is a slight height change when toggling preview on/off. see video

https://user-images.githubusercontent.com/52937121/168894608-9e83d61c-06c9-4834-953f-1a2515a90c7f.mov

**Comment by michelengelen** [comment]
>     1. Unable to post by hitting Enter while in preview mode

fixed!

>     2. If preview is ON and user hides the formatting bar, and we can't post using Enter, the experience is slightly broken as it's not visually apparent that the preview is enabled

fixed by re-arranging the icons. @matthewbirtch would that be ok for you as it is now?

>     3. Able to add emoji to draft post in preview mode

fixed, but it disables the preview mode by default.

>     4. ~Able to remove attachments in preview mode~ discussed with Matt, we can leave this as is

OK

>     5. 3-dot menu is disconnected from `...` button in narrow widths

Fixed!

>     6. 3-dot menu opacity is not correct when preview mode is active

Fixed

>     7. Should we close 3-dot menu on outside click? Currently user has to click on the 3-dot button again to close, can't be dismissed in any other way

Yes, IMO we should! Will look into it later

>     8. When in preview mode and clicking on formatting options buttons, a long string can be inadvertently added to the back of the draft text, user can accidentally miss any changes being added on preview screen

How should we handle this? IMO we have 2 options:
1. disable all editing when in preview mode
2. disable preview mode once the editing buttons are used

> 9. The send button doesn't show/hide dynamically when changing the window size.

Works for me!

**Comment by michelengelen** [comment]
>     1. 3-dot menu tucks under file icon (preview off)

Fixed!

>     2. As @matthewbirtch previously mentioned, there is a slight height change when toggling preview on/off. see video

This is due to an emoji being present and not easily fixed since the line-height will change automatically when previewing.  I will try and have a go at this anyways, but be aware that this might not be possible.

**Comment by michelengelen** [comment]
@jgilliam17 See comments above!

**Comment by michelengelen** [comment]
cc @matthewbirtch

**Comment by michelengelen** [comment]
@jgilliam17 and @matthewbirtch I fixed all of the above findings and added an outside click listener to the hidden controls popover, so we can close it by clicking away from it.

There is still a bug i am currently looking into that when you have RHS and center channel inputs open, without content and center channel input focussed pressing shift will move the focus to the RHS. If I cannot find and eliminate the cause i will open up a bug ticket for that

**Comment by michelengelen** [comment]
> > ```
> > 2. As @matthewbirtch previously mentioned, there is a slight height change when toggling preview on/off. see video
> > ```
> 
> This is due to an emoji being present and not easily fixed since the line-height will change automatically when previewing. I will try and have a go at this anyways, but be aware that this might not be possible.

As I expected the slight height change comes from replacing the emoji code with an actual emoji. This is not feasible to fix, since all other markdowns will also result in an even bigger height change.

> 8. When in preview mode and clicking on formatting options buttons, a long string can be inadvertently added to the back of the draft text, user can accidentally miss any changes being added on preview screen

As for this: I have disabled all markdown formatting options in preview mode.

**Comment by matthewbirtch** [comment]
@michelengelen not sure what happened, but the background on the formatting bar is somehow using the sidebar bg color. 
image

**Comment by matthewbirtch** [comment]
> There is still a bug i am currently looking into that when you have RHS and center channel inputs open, without content and center channel input focussed pressing shift will move the focus to the RHS. If I cannot find and eliminate the cause i will open up a bug ticket for that

@michelengelen this should be fixed in this PR as it prevents users typing a capital letter in the message input and we can't merge it like that. Most messages start with a capital letter, so this is a blocking bug.

**Comment by matthewbirtch** [comment]
> 3. Able to add emoji to draft post in preview mode

This is still not fixed. In preview mode, the emoji picker should be disabled just like the other formatting buttons and controls

**Comment by matthewbirtch** [comment]
> > ```
> > 8. When in preview mode and clicking on formatting options buttons, a long string can be inadvertently added to the back of the draft text, user can accidentally miss any changes being added on preview screen
> > ```
> 
> How should we handle this? IMO we have 2 options:
> 
> 1. disable all editing when in preview mode
> 
> 2. disable preview mode once the editing buttons are used

All formatting controls, including the emoji button should be disabled while in preview mode.

**Comment by michelengelen** [comment]
> > 3. Able to add emoji to draft post in preview mode
> 
> This is still not fixed. In preview mode, the emoji picker should be disabled just like the other formatting buttons and controls

I thought that was intentional ... ok, will fix it!

**Comment by jgilliam17** [comment]
Thanks @michelengelen and @AshishDhama 
Apologies, missed this on my first pass, adding notes from testing on Ubuntu and Win10.
1. Preview icon tooltip on Win and Ubuntu, shows `Emoji/Gif picker`
2. Keyboard shortcut conflict `ctrl` `shift` `7` , `ctrl` `shift` `8` and `ctrl` `shift` `9` are used for servers - I may be in minority here with this many servers on the desktop app 😄 , this makes the shortcuts unusable in both desktop and browsers - nothing happens in Ubuntu and Win10.
3. With file attachment, turning formatting bar off leaves quite a bit of white space where the bar used to be
4. Opening RHS, does not truncate formatting bar, causes overlap of icons (all OSes)
5. nit - I think we should update tooltip text to title case ``Emoji/Gif picker` to `Emoji/Gif Picker`, Upload files" to `Upload Files`

**Comment by jgilliam17** [comment]
@michelengelen 
Formatting bar is themed correctly now, but hidden options background still has the sidebar color 
Screenshot 2022-05-18 122243

**Comment by michelengelen** [comment]
> @michelengelen Formatting bar is themed correctly now, but hidden options background still has the sidebar color Screenshot 2022-05-18 122243

I changed the background to take the color used in the center-channel background. This will be lighter than the color for the formattingBar. We cannot use the same color as with it, since it uses an opacity and would "shine-through", but we can think of a better option for that after the implementation.

Is that ok for you?

**Comment by Willyfrog** [comment]
@nevyangelova @michelengelen @matthewbirtch can we have a go, we'll be adressing the feedback in subsequent PRs

**Review by Willyfrog — APPROVED** [review]

**Comment by matthewbirtch** [comment]
> @nevyangelova @michelengelen @matthewbirtch can we have a go, we'll be adressing the feedback in subsequent PRs

That's fine, @Willyfrog are we tracking all the remaining issues somewhere?

**Comment by matthewbirtch** [comment]
@michelengelen when you re-arranged the controls, you moved the attachment icon in to the formatting bar. So now you can only access the attachment icon if you open the formatting bar. I think it's more important to have the attachment available rather than the preview icon when the formatting bar is closed. Can we track this as another follow-up?

**Review by matthewbirtch — APPROVED** [review]
Approving as long as all other issues noted can be resolved in follow-ups.

**Comment by Willyfrog** [comment]
> That's fine, @Willyfrog are we tracking all the remaining issues somewhere?

We'll Be adding them as bugs to the epic in Jira

**Review by michelengelen — APPROVED** [review]

**Review by nevyangelova — APPROVED** [review]

**Comment by jgilliam17** [comment]
Thanks @AshishDhama and @michelengelen 
Tested, looks good to merge - follow up tickets for fixing e2es and resolving other issues mentioned here are filed and linked below.
https://mattermost.atlassian.net/browse/MM-44391
https://mattermost.atlassian.net/browse/MM-44390
https://mattermost.atlassian.net/browse/MM-44389
https://mattermost.atlassian.net/browse/MM-44392
https://mattermost.atlassian.net/browse/MM-44393
https://mattermost.atlassian.net/browse/MM-44394

**Review by jgilliam17 — APPROVED** [review]

**Pull request merged**
